src/main/java/org/ice4j/ice/Component.java

```diff
@@ -1207,17 +1212,53 @@ public Logger getLogger()
     public void send(byte[] buffer, int offset, int length)
             throws IOException
     {
-        CandidatePair pair = getSelectedPair();
+        CandidatePair pair = findPair(lastReceivedFrom);
         if (pair == null)
         {
-            logger.debug("No selected pair, will try valid for sending");
-            pair = parentStream.getValidPair(this);
-            if (pair == null)
+            throw new IOException("No valid pair.");
+        }
+
+        AddressAndSocket addressAndSocket = getAddressAndSocket(pair);
+        if (addressAndSocket == null)
+        {
+            throw new IOException("No valid socket.");
+        }
+
+        DatagramPacket p = new DatagramPacket(buffer, offset, length);
+        p.setSocketAddress(addressAndSocket.remoteAddress);
+        addressAndSocket.socket.send(p);
+    }
+
+    private CandidatePair findPair(SocketAddress remoteAddress)
+    {
+        CandidatePair pair = null;
+        if (AgentConfig.config.getSendToLastReceivedFromAddress() && remoteAddress != null)
```

Comment:
This searches all the keep alive pairs on every packet, which is inefficient for the common case.  Is there a way to fast-path the common case where this is unchanged?  Perhaps checking and setting a flag on receive?